### DISCUSSION TITLE
Testing Pagerdutyzabbix

### DISCUSSION THREAD

**Discussion by ichi-the-one** [sonic-com/pagerduty2zabbix#12] [Q&A] [answered]
Hello Eric. Thank you for sharing your contribution. The use case your handling here is very usueful and I wanted to test in my PoC setup, but didn't succeed in doing so. Would you plz give me more insight about the deployment.I'll be very gratefull to you

I specifically didn't understand the part below, although ;my zabbix-->pagerduty works just fine

_I recommend setting the token in the Media type to {ALERT.SENDTO} and putting your PagerDuty API token into "Send to" of the user's media configuration. (so you have the easy option of additional PD integrations for different teams, etc)_

Also, do we need to open specific ports on the server where we deployed the script(I deployed it to zabbix frontend). what are the ownership of the script?. 
in step 10 for setting pagerduty, what url you're refering to?  is it https://zabbix/cgi-bin/pd2zabbix.cgi ?
I basically went through all the step you gave on teh readme, but still could'nt get the integration to work. Also I'm unable to see any logs in the https/error_log.

**Comment by eric-eisenhart** [comment, marked as answer]
Hello @ichi-the-one.

I'll try to answer what I can...

> I specifically didn't understand the part below, although ;my zabbix-->pagerduty works just fine
> 
> _I recommend setting the token in the Media type to {ALERT.SENDTO} and putting your PagerDuty API token into "Send to" of the user's media configuration. (so you have the easy option of additional PD integrations for different teams, etc)_

If you have it working, and don't intend to support multiple different groups in your Zabbix setup, it's really not important.

In Zabbix / Alerts / Media Types / PagerDuty, for the "token" parameter, instead of the token hard-coded there, I have it set to `{ALERT.SENDTO}`.

For the user that's set up to send to PagerDuty, in the "PagerDuty" media entry, I have the PagerDuty api token in the "Send to" field.

That gets the contents of the user "send to" field

And instead of naming that user "PagerDuty" or something like that, I have it named after the group who the alerts are for.

Doing things that way means that it's easy for me to have specific alerts go to other PagerDuty services/automations.

> Also, do we need to open specific ports on the server where we deployed the script(I deployed it to zabbix frontend). 

Yes, you'll need the https or http port open to at least the PagerDuty WebHook IPs: https://developer.pagerduty.com/docs/9a349b09b87b7-webhook-i-ps

Ours is open to the world, but is also running on a separate server from Zabbix itself.

> what are the ownership of the script?. 

On mine the script is owned by root, but usually doesn't matter, as long as the web server (httpd) user can read and execute the script.

On a RHEL/Rocky server running apache httpd and the fairly stock Zabbix config, I was able to set it up on my zabbix server with:
```bash
cd /var/www/cgi-bin
git clone https://github.com/sonic-com/pagerduty2zabbix.git
cd pagerduty2zabbix
cp pagerduty2zabbix.conf.example pagerduty2zabbix.conf
vim pagerduty2zabbix.conf # edit with your favorite editor
```

> in step 10 for setting pagerduty, what url you're refering to? is it https://zabbix/cgi-bin/pd2zabbix.cgi ? 

It will need to be the fully-qualified URL that PagerDuty's servers can use. A shortname like "zabbix" definitely won't work from the servers the WebHooks run from.

In my config, that would be something like https://zabbix.example.com/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi

You can try it with a web browser or `curl` to verify...

If you do a simple `curl https://zabbix.example.com/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi` you should see output that starts with something like this:
```html
<h1>Software error:</h1>
<pre>No json_payload from webhook POSTDATA
</pre>
```

> I basically went through all the step you gave on teh readme, but still could'nt get the integration to work. Also I'm unable to see any logs in the https/error_log.

Set debug in the config to 3 or higher until you have it working to see more of what's going on...

Watch the access logs too. If you have the firewall stuff and PagerDuty and everything set up appropriately, you should be able to "Test" the webhook and have a "ping" sent to it, which will show up in your access logs as PagerDuty doing a POST to the URL.

**Comment by eric-eisenhart** [comment]
I tried to update README.md in f0fc083 to make some of that a bit clearer...

**Comment by eric-eisenhart** [comment]
Screenshots of the PD webhook configuration that I meant:

![image](https://github.com/sonic-com/pagerduty2zabbix/assets/1577595/6a026298-48b3-45e2-a8e5-da9389baa405)

![image](https://github.com/sonic-com/pagerduty2zabbix/assets/1577595/20b0e0df-6a11-4307-ab52-42a03906daec)

**Comment by ichi-the-one** [comment]
Amazing. Thank you so much for your detailed answer.

**Reply by eric-eisenhart** [reply]
@ichi-the-one Please report back whether or not you got it all working.

And thank you for your assistance in making this easier on future other people. :)

**Comment by ichi-the-one** [comment]
It's way better than yesterday. I can see the logs while testing. I have this error while trying to update a ticket. 

[Wed Feb 07 21:49:43.910534 2024] [cgid:error] [pid 26640:tid 26821] [client 54.213.187.133:29801] [Wed Feb  7 21:49:43 2024] pd2zabbix.cgi: Unable to fetch alerts from PagerDuty: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 21:49:43.910882 2024] [cgid:error] [pid 26640:tid 26821] [client 54.213.187.133:29801] [Wed Feb  7 21:49:43 2024] pd2zabbix.cgi: 500 Can't connect to api.pagerduty.com:443 (Permission denied): /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 21:49:43.910980 2024] [cgid:error] [pid 26640:tid 26821] [client 54.213.187.133:29801] [Wed Feb  7 21:49:43 2024] pd2zabbix.cgi: Content-Type: text/plain: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 21:49:43.911086 2024] [cgid:error] [pid 26640:tid 26821] [client 54.213.187.133:29801] [Wed Feb  7 21:49:43 2024] pd2zabbix.cgi: Client-Date: Wed, 07 Feb 2024 21:49:43 GMT: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 21:49:43.911182 2024] [cgid:error] [pid 26640:tid 26821] [client 54.213.187.133:29801] [Wed Feb  7 21:49:43 2024] pd2zabbix.cgi: Client-Warning: Internal response: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 21:49:43.911281 2024] [cgid:error] [pid 26640:tid 26821] [client 54.213.187.133:29801] [Wed Feb  7 21:49:43 2024] pd2zabbix.cgi: : /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 21:49:43.911510 2024] [cgid:error] [pid 26640:tid 26821] [client 54.213.187.133:29801] [Wed Feb  7 21:49:43 2024] pd2zabbix.cgi: Can't connect to api.pagerduty.com:443 (Permission denied): /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 21:49:43.911587 2024] [cgid:error] [pid 26640:tid 26821] [client 54.213.187.133:29801] [Wed Feb  7 21:49:43 2024] pd2zabbix.cgi: : /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 21:49:43.911727 2024] [cgid:error] [pid 26640:tid 26821] [client 54.213.187.133:29801] [Wed Feb  7 21:49:43 2024] pd2zabbix.cgi: Permission denied at /usr/share/perl5/vendor_perl/LWP/Protocol/http.pm line 50.: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 21:49:43.911796 2024] [cgid:error] [pid 26640:tid 26821] [client 54.213.187.133:29801] [Wed Feb  7 21:49:43 2024] pd2zabbix.cgi: : /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 21:49:43.911922 2024] [cgid:error] [pid 26640:tid 26821] [client 54.213.187.133:29801] [Wed Feb  7 21:49:43 2024] pd2zabbix.cgi: Died at /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi line 420.: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi

**Reply by eric-eisenhart** [reply]
> [Wed Feb 07 21:49:43.910534 2024] [cgid:error] [pid 26640:tid 26821] [client 54.213.187.133:29801] [Wed Feb 7 21:49:43 2024] pd2zabbix.cgi: Unable to fetch alerts from PagerDuty: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi 
> [Wed Feb 07 21:49:43.910882 2024] [cgid:error] [pid 26640:tid 26821] [client 54.213.187.133:29801] [Wed Feb 7 21:49:43 2024] pd2zabbix.cgi: 500 Can't connect to api.pagerduty.com:443 (Permission denied): /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi 

I think that might mean you've got some firewall or selinux policy blocking the ability for that server to connect to https://api.pagerduty.com/.

Can you `curl -v https://api.pagerduty.com/` from the command-line of the server? (should give you a "302 Moved" / "302 Found" type response)

I think it's less likely (should be a "401" error, not a "500" error, but could mean that your `pdtoken` in the config isn't right. My pdtoken is a "User Token" generated from "My Profile" / "User Settings"; but as far as I know a [General Access REST API Token](https://support.pagerduty.com/docs/api-access-keys#section-generating-a-general-access-rest-api-key) will just as well, and can be explicitly read-only.

What do you have debug= set to? Turning that to a higher value might help get more details about what's going on.

**Reply by ichi-the-one** [reply]
U've guessed it right. in my case it was Selinux.

**Comment by ichi-the-one** [comment]
when I use the token from the service in pagerduty and put it as the value of pdtoken I have this 


[Wed Feb 07 22:24:33.921306 2024] [cgid:error] [pid 27951:tid 28105] [client 54.213.187.133:39491] [Wed Feb  7 22:24:33 2024] pd2zabbix.cgi: Unable to fetch alerts from PagerDuty: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 22:24:33.921564 2024] [cgid:error] [pid 27951:tid 28105] [client 54.213.187.133:39491] [Wed Feb  7 22:24:33 2024] pd2zabbix.cgi: HTTP/1.1 401 Unauthorized: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 22:24:33.921826 2024] [cgid:error] [pid 27951:tid 28105] [client 54.213.187.133:39491] [Wed Feb  7 22:24:33 2024] pd2zabbix.cgi: Connection: close: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 22:24:33.921991 2024] [cgid:error] [pid 27951:tid 28105] [client 54.213.187.133:39491] [Wed Feb  7 22:24:33 2024] pd2zabbix.cgi: Date: Wed, 07 Feb 2024 22:24:33 GMT: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 22:24:33.922100 2024] [cgid:error] [pid 27951:tid 28105] [client 54.213.187.133:39491] [Wed Feb  7 22:24:33 2024] pd2zabbix.cgi: Server: nginx: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 22:24:33.922249 2024] [cgid:error] [pid 27951:tid 28105] [client 54.213.187.133:39491] [Wed Feb  7 22:24:33 2024] pd2zabbix.cgi: Content-Length: 0: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 22:24:33.922376 2024] [cgid:error] [pid 27951:tid 28105] [client 54.213.187.133:39491] [Wed Feb  7 22:24:33 2024] pd2zabbix.cgi: Access-Control-Allow-Origin: *: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 22:24:33.922517 2024] [cgid:error] [pid 27951:tid 28105] [client 54.213.187.133:39491] [Wed Feb  7 22:24:33 2024] pd2zabbix.cgi: Client-Date: Wed, 07 Feb 2024 22:24:33 GMT: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 22:24:33.922662 2024] [cgid:error] [pid 27951:tid 28105] [client 54.213.187.133:39491] [Wed Feb  7 22:24:33 2024] pd2zabbix.cgi: Client-Peer: 35.163.146.4:443: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 22:24:33.922972 2024] [cgid:error] [pid 27951:tid 28105] [client 54.213.187.133:39491] [Wed Feb  7 22:24:33 2024] pd2zabbix.cgi: Client-Response-Num: 1: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 22:24:33.923133 2024] [cgid:error] [pid 27951:tid 28105] [client 54.213.187.133:39491] [Wed Feb  7 22:24:33 2024] pd2zabbix.cgi: Client-SSL-Cert-Issuer: /C=US/O=DigiCert Inc/CN=DigiCert TLS RSA SHA256 2020 CA1: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 22:24:33.923326 2024] [cgid:error] [pid 27951:tid 28105] [client 54.213.187.133:39491] [Wed Feb  7 22:24:33 2024] pd2zabbix.cgi: Client-SSL-Cert-Subject: /C=US/ST=California/L=San Francisco/O=PagerDuty, Inc./CN=*.pagerduty.com: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 22:24:33.923468 2024] [cgid:error] [pid 27951:tid 28105] [client 54.213.187.133:39491] [Wed Feb  7 22:24:33 2024] pd2zabbix.cgi: Client-SSL-Cipher: ECDHE-RSA-AES128-GCM-SHA256: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 22:24:33.923583 2024] [cgid:error] [pid 27951:tid 28105] [client 54.213.187.133:39491] [Wed Feb  7 22:24:33 2024] pd2zabbix.cgi: Client-SSL-Socket-Class: IO::Socket::SSL: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 22:24:33.923700 2024] [cgid:error] [pid 27951:tid 28105] [client 54.213.187.133:39491] [Wed Feb  7 22:24:33 2024] pd2zabbix.cgi: Client-SSL-Version: TLSv1_2: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 22:24:33.923826 2024] [cgid:error] [pid 27951:tid 28105] [client 54.213.187.133:39491] [Wed Feb  7 22:24:33 2024] pd2zabbix.cgi: Client-Warning: Missing Authenticate header: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 22:24:33.923963 2024] [cgid:error] [pid 27951:tid 28105] [client 54.213.187.133:39491] [Wed Feb  7 22:24:33 2024] pd2zabbix.cgi: X-Request-Id: 6928809703a5747f6aefe12af3a75c95: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 22:24:33.924059 2024] [cgid:error] [pid 27951:tid 28105] [client 54.213.187.133:39491] [Wed Feb  7 22:24:33 2024] pd2zabbix.cgi: : /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 22:24:33.924134 2024] [cgid:error] [pid 27951:tid 28105] [client 54.213.187.133:39491] [Wed Feb  7 22:24:33 2024] pd2zabbix.cgi: : /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi
[Wed Feb 07 22:24:33.924288 2024] [cgid:error] [pid 27951:tid 28105] [client 54.213.187.133:39491] [Wed Feb  7 22:24:33 2024] pd2zabbix.cgi: Died at /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi line 420.: /var/www/cgi-bin/pagerduty2zabbix/pd2zabbix.cgi



when I use instead a user token and use it in pdtoken as you mentionned in the readme:     Get an API token from PagerDuty that can update the relevant PagerDuty events and set pdtoken to that. (profile pic > User Settings > Create API User Token)  +++> then I try to update the ticket , I have absolutely no trace coming in myserver

**Reply by eric-eisenhart** [reply]
It needs to be the "User API Token" or maybe the "Generic API Token", not a "Service" token.

> then I try to update the ticket , I have absolutely no trace coming in myserver

If debug=0, and everything is working, it shouldn't log anything to the error log. So...  That looks like it could be working...

If you Acknowledge an alert in PagerDuty that came from Zabbix, does the Zabbix event/problem get acknowledged? Or add a note to the alert in PD, does the zabbix alert get a comment?

If you look at the list of "problems" in Zabbix, you can click on the little speech ballon (💬) or the arrow with number to see some details, and in a default pd2zabbix config that should include a comment with a link to the alert in PagerDuty: 
![image](https://github.com/sonic-com/pagerduty2zabbix/assets/1577595/0cc616b7-d49c-4dbc-a4fa-644e0cd52d23)
![image](https://github.com/sonic-com/pagerduty2zabbix/assets/1577595/62df4e2c-862b-4569-b1f5-1908a0588125)
![image](https://github.com/sonic-com/pagerduty2zabbix/assets/1577595/f32d0424-1add-4993-a3bd-32d49fce67f6)

**Comment by ichi-the-one** [comment]
I confirm. it's working. Zabbix is getting updated according to PD. Thank you so much.

**Reply by eric-eisenhart** [reply]
Awesome

**Reply by ichi-the-one** [reply]
Thank you so much for your help
